File: src/air_mouse_cpp/src/failsafe.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/bool.hpp>
#include <std_msgs/msg/string.hpp>
#include <sensor_msgs/msg/laser_scan.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <px4_msgs/msg/battery_status.hpp>
#include <chrono>

class FailsafeNode : public rclcpp::Node {
public:
  FailsafeNode() : Node("failsafe_node") {
    this->declare_parameter("max_height", 2.44);
    this->declare_parameter("arena_size", 15.0);
    this->declare_parameter("vio_timeout", 2.0);
    this->declare_parameter("lidar_timeout", 2.0);
    this->declare_parameter("low_battery", 14.0);

    max_h_=this->get_parameter("max_height").as_double();
    arena_=this->get_parameter("arena_size").as_double();
    vio_timeout_=this->get_parameter("vio_timeout").as_double();
    lidar_timeout_=this->get_parameter("lidar_timeout").as_double();
    low_batt_=this->get_parameter("low_battery").as_double();

    pose_sub_=this->create_subscription<geometry_msgs::msg::PoseStamped>("/offboard/pose", 10, std::bind(&FailsafeNode::poseCb, this, std::placeholders::_1));
    scan_sub_=this->create_subscription<sensor_msgs::msg::LaserScan>("/scan", 10, std::bind(&FailsafeNode::scanCb, this, std::placeholders::_1));
    vio_sub_=this->create_subscription<geometry_msgs::msg::PoseStamped>("/vins_estimator/camera_pose", 10, std::bind(&FailsafeNode::vioCb, this, std::placeholders::_1));
    e_stop_sub_=this->create_subscription<std_msgs::msg::Bool>("/emergency_stop", 10, std::bind(&FailsafeNode::eStopCb, this, std::placeholders::_1));
    batt_sub_=this->create_subscription<px4_msgs::msg::BatteryStatus>("/fmu/out/battery_status", 10, std::bind(&FailsafeNode::battCb, this, std::placeholders::_1));

    abort_pub_=this->create_publisher<std_msgs::msg::Bool>("/mission/abort", 10);
    reason_pub_=this->create_publisher<std_msgs::msg::String>("/failsafe/reason", 10);
    status_pub_=this->create_publisher<std_msgs::msg::String>("/failsafe/status", 10);

    last_vio_=this->now();
    last_lidar_=this->now();

    timer_=this->create_wall_timer(std::chrono::milliseconds(500), std::bind(&FailsafeNode::check, this));
    RCLCPP_INFO(this->get_logger(), "C++ Failsafe max_h %.2f arena %.1f", max_h_, arena_);
  }

private:
  void poseCb(const geometry_msgs::msg::PoseStamped::SharedPtr msg) { pose_=*msg; has_pose_=true; }
  void scanCb(const sensor_msgs::msg::LaserScan::SharedPtr) { last_lidar_=this->now(); }
  void vioCb(const geometry_msgs::msg::PoseStamped::SharedPtr) { last_vio_=this->now(); }
  void eStopCb(const std_msgs::msg::Bool::SharedPtr msg) { if (msg->data) trigger("Emergency stop"); }
  void battCb(const px4_msgs::msg::BatteryStatus::SharedPtr msg) { voltage_=msg->voltage_v; }

  void check() {
    std::string reason;
    if (emergency_) reason="Emergency stop";
    if (has_pose_ && pose_.pose.position.z > max_h_) reason="Height limit "+std::to_string(pose_.pose.position.z)+" > "+std::to_string(max_h_);
    if (has_pose_) {
      double x=pose_.pose.position.x, y=pose_.pose.position.y;
      if (std::abs(x)>arena_/2 || std::abs(y)>arena_/2) reason="Geofence breach";
    }
    if ((this->now()-last_vio_).seconds() > vio_timeout_) reason="VIO timeout";
    if ((this->now()-last_lidar_).seconds() > lidar_timeout_) reason="Lidar timeout";
    if (voltage_>0 && voltage_ < low_batt_) reason="Low battery "+std::to_string(voltage_);

    if (!reason.empty()) trigger(reason);
    else {
      std_msgs::msg::String s; s.data="OK"; status_pub_->publish(s);
    }
  }

  void trigger(const std::string& reason) {
    RCLCPP_ERROR(this->get_logger(), "FAILSAFE: %s", reason.c_str());
    std_msgs::msg::String r; r.data=reason; reason_pub_->publish(r);
    std_msgs::msg::Bool b; b.data=true; abort_pub_->publish(b);
  }

  double max_h_, arena_, vio_timeout_, lidar_timeout_, low_batt_, voltage_=16.8;
  geometry_msgs::msg::PoseStamped pose_;
  bool has_pose_=false, emergency_=false;
  rclcpp::Time last_vio_, last_lidar_;
  rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr pose_sub_, vio_sub_;
  rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr scan_sub_;
  rclcpp::Subscription<std_msgs::msg::Bool>::SharedPtr e_stop_sub_;
  rclcpp::Subscription<px4_msgs::msg::BatteryStatus>::SharedPtr batt_sub_;
  rclcpp::Publisher<std_msgs::msg::Bool>::SharedPtr abort_pub_;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr reason_pub_, status_pub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<FailsafeNode>());
  rclcpp::shutdown();
  return 0;
}

File: src/air_mouse_mission/air_mouse_mission/failsafe.py
#!/usr/bin/env python3
"""
failsafe.py - NIDAR AirMouse failsafe + emergency stop
Requirements from NIDAR:
- Must include emergency stop/mission abort capability and failsafe features
- Corridor height clearance 8 feet (2.44m) at all times -> enforce max height
- Max mission 30 mins
- Return to home if comms lost, low battery, VIO failure, lidar failure

This node monitors:
- /emergency_stop (Bool) from ground station button
- Battery voltage (from PX4)
- VIO health (/vins_estimator/odometry timeout)
- Lidar health (/scan timeout)
- Height limit (z > 2.44m -> force down)
- Geofence (15x15m arena)
- RC loss

Publishes:
- /mission/abort (Bool)
- /failsafe/reason (String)
- Triggers LAND via VehicleCommand
"""
import rclpy
from rclpy.node import Node
from std_msgs.msg import Bool, String, Float32
from sensor_msgs.msg import LaserScan
from geometry_msgs.msg import PoseStamped
from px4_msgs.msg import VehicleStatus, BatteryStatus
import time

class FailsafeNode(Node):
    def __init__(self):
        super().__init__('failsafe_node')
        self.declare_parameter('max_height', 2.44)  # 8 feet
        self.declare_parameter('arena_size', 15.0)
        self.declare_parameter('vio_timeout', 2.0)
        self.declare_parameter('lidar_timeout', 2.0)
        self.declare_parameter('low_battery', 14.0)  # 4S

        self.max_height = self.get_parameter('max_height').value
        self.arena_size = self.get_parameter('arena_size').value
        self.vio_timeout = self.get_parameter('vio_timeout').value
        self.lidar_timeout = self.get_parameter('lidar_timeout').value

        self.last_vio_time = time.time()
        self.last_lidar_time = time.time()
        self.current_pose = None
        self.battery_voltage = 16.8
        self.emergency_stop = False

        # Subs
        self.pose_sub = self.create_subscription(PoseStamped, '/offboard/pose', self.pose_callback, 10)
        self.scan_sub = self.create_subscription(LaserScan, '/scan', self.scan_callback, 10)
        self.e_stop_sub = self.create_subscription(Bool, '/emergency_stop', self.e_stop_callback, 10)
        self.vio_sub = self.create_subscription(PoseStamped, '/vins_estimator/camera_pose', self.vio_callback, 10)
        self.battery_sub = self.create_subscription(BatteryStatus, '/fmu/out/battery_status', self.battery_callback, 10)

        # Pubs
        self.abort_pub = self.create_publisher(Bool, '/mission/abort', 10)
        self.reason_pub = self.create_publisher(String, '/failsafe/reason', 10)
        self.status_pub = self.create_publisher(String, '/failsafe/status', 10)

        self.timer = self.create_timer(0.5, self.check_failsafe)

        self.get_logger().info(f"Failsafe: max height {self.max_height}m, arena {self.arena_size}m, vio timeout {self.vio_timeout}s")

    def pose_callback(self, msg):
        self.current_pose = msg

    def scan_callback(self, msg):
        self.last_lidar_time = time.time()

    def vio_callback(self, msg):
        self.last_vio_time = time.time()

    def battery_callback(self, msg):
        # PX4 BatteryStatus voltage
        self.battery_voltage = msg.voltage_v

    def e_stop_callback(self, msg):
        self.emergency_stop = msg.data
        if self.emergency_stop:
            self.trigger_abort("Emergency stop button pressed")

    def check_failsafe(self):
        reason = None

        # Check emergency stop
        if self.emergency_stop:
            reason = "Emergency stop"

        # Check height
        if self.current_pose and self.current_pose.pose.position.z > self.max_height:
            reason = f"Height limit exceeded: {self.current_pose.pose.position.z:.2f} > {self.max_height}"

        # Check geofence
        if self.current_pose:
            x = self.current_pose.pose.position.x
            y = self.current_pose.pose.position.y
            if abs(x) > self.arena_size/2 or abs(y) > self.arena_size/2:
                reason = f"Geofence breach: ({x:.1f}, {y:.1f}) outside {self.arena_size}m"

        # Check VIO timeout
        if time.time() - self.last_vio_time > self.vio_timeout:
            reason = f"VIO timeout: {time.time()-self.last_vio_time:.1f}s"

        # Check lidar timeout
        if time.time() - self.last_lidar_time > self.lidar_timeout:
            reason = f"Lidar timeout: {time.time()-self.last_lidar_time:.1f}s"

        # Check battery
        if self.battery_voltage < self.get_parameter('low_battery').value:
            reason = f"Low battery: {self.battery_voltage:.1f}V"

        if reason:
            self.trigger_abort(reason)
        else:
            # Publish OK status
            status = String()
            status.data = "OK"
            self.status_pub.publish(status)

    def trigger_abort(self, reason):
        self.get_logger().error(f"FAILSAFE TRIGGERED: {reason}")

        reason_msg = String()
        reason_msg.data = reason
        self.reason_pub.publish(reason_msg)

        abort_msg = Bool()
        abort_msg.data = True
        self.abort_pub.publish(abort_msg)

def main(args=None):
    rclpy.init(args=args)
    node = FailsafeNode()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
